promise.go

```diff
@@ -0,0 +1,88 @@
+package format
+
+import (
+	"context"
+	"sync"
+)
+
+// TODO: I renamed this to NodePromise because:
+// 1. NodeGetter is a naming conflict.
+// 2. It's a promise...
+
+// TODO: Should this even be an interface? It seems like a simple struct would
```

Comment:
This could probably be a simple struct as opposed to an interface. I'm not sure i can see a scenario where we would want to implement this differently